Initial PR

bjackman opened the pull request:
I created this repository without any code in it, now I'm creating the first commit as a PR, hopefully this gives us an easy way to review the code.

bjackman commented:
I just remembered I'd initially planned on having a commit that copy-pastes the stdlib code and another commit that retrofits the async bits. I can amend it to include that if you folks think it's useful?

bjackman committed:
Copy filepath.Glob code from Go stand library

Copied from
 - https://golang.org/src/path/filepath/match.go
 - https://golang.org/src/path/filepath/match_test.go

bjackman committed:
Modify copied code so it can breathe outside of the stdlib

- Update references to public parts of the standard library to be qualified by
  the package name
- Replace private volumeNameLen with len(filepath.VolumeName())
- Update name of the package of this code
- Tweak some tests so that they work without assumptions about the source
  structure (added testdata/ directory).
- Add updated copyright, licenses, and other repo boilerplate.

gnoack commented on streaming.go:
```diff
@@ -0,0 +1,239 @@
+// stream finds files matching pattern and sends their paths on the results
+// channel. It stops (returning nil) if the cancel channel is closed.
+// The caller must drain the results channel.
+func stream(pattern string, results chan<- string, cancel <-chan struct{}) (err error) {
```
https://golang.org/doc/effective_go.html#named-results

I can't find this in the style guide in that phrasing, but I agree in principle–naming these has done more harm than good for me when I've used it. `foo, err := bar()` looks like it redefines `err`, but it doesn't if `err` is a named result parameter.

bjackman replied:
Ack but that would introduce unnecessary diff from the code this is copied from (it was probably written before this Go norm developed). I'm splitting the commits up so we can see which code is introduced by this repo.

bjackman replied:
Sorry, there was a race condition, I wrote my comment in response to Adam's original comment.

bjackman committed:
Remove Match and ErrBadPattern

We'll just use the ones from the stdlib

bjackman committed:
Add the actual streaming funcionality

This is where this repo justifies its existence - the ancestor commits have just
been preparation to give this commit a nice helpful diff. See the README and doc
strings for an explanation of what this is all about.

bjackman commented on streaming.go:
```diff
@@ -0,0 +1,239 @@
+
+// GlobSync is like filepath.Glob except
+//
+// - It can be canceled via the context
```
Thanks, you're right. Changing it to just be a para.

bjackman commented on streaming_test.go:
```diff
@@ -0,0 +1,360 @@
+	if err != nil {
+		t.Fatalf("Glob error for %q: %s", pattern, err)
+	}
+	if !reflect.DeepEqual(matches, want) {
```
Ack, but this is also not my code

bjackman commented on streaming_test.go:
```diff
@@ -0,0 +1,360 @@
+		t.Fatalf("Glob error for %q: %s", pattern, err)
+	}
+	if !reflect.DeepEqual(matches, want) {
+		t.Fatalf("Glob(%#q) = %v want %v", pattern, matches, want)
```
Ditto.. I tried it and it does actually work :shrug:

bjackman commented on streaming.go:
```diff
@@ -0,0 +1,239 @@
+//   allows it to run in O(d+m) memory and O(n) time, where m is the number of
+//   match results, d is the depth of the directory tree the pattern is
+//   concerned with, and n is the number of files in that tree.
+func GlobSync(ctx context.Context, pattern string) ([]string, error) {
```
OK renamed the package, the functions will now be `glob.Glob` and `glob.Stream`.

bjackman commented on streaming.go:
```diff
@@ -0,0 +1,239 @@
+		if err := stream(pattern, g.results, ctx.Done()); err != nil {
+			g.errors <- err
+		}
+		close(g.results)
```
Nice, thanks

bjackman commented on streaming.go:
```diff
@@ -0,0 +1,239 @@
+	}
+	go func() {
+		if err := stream(pattern, g.results, ctx.Done()); err != nil {
+			g.errors <- err
```
Thanks, fixed

bjackman commented on streaming.go:
```diff
@@ -0,0 +1,239 @@
+		case <-cancel:
+			return nil
+		default:
+		}
```
Hmm true. I've added the `cancel` check in a select with the send on `results` but I think we also need the separate `cancel` check so we are responsive to canceling even when we're just grinding through negative results.

bjackman reviewed (commented):
Thanks for the review!

bjackman commented:
Er, top-level comments turn out not to be threaded, but my comment above about separating it into separate commits is now satisfied.

PTAL, thanks!

bjackman commented on glob.go:
```diff
@@ -12,30 +12,115 @@
+
+type GlobResult struct {
+	errors  chan error
+	results chan string
```
Yeah I thought about that, but then the issue is we have to do one of:

- `stream` has to send its error down a channel instead of returning it. I actually tried writing this out and it actually looked even worse than it sounds
- `stream` still gets a `chan string` and returns its error, and then the caller wraps up all the `string`s into a `struct {string, error} {s, nil}` and forwards them, then sends `struct {string, error}{"", err}` after `stream` has returned. This doesn't seem totally insane but I'm not really sure if it overall reduces the cognitive load...

Any thoughts?

gnoack commented on LICENSE:
```diff
@@ -0,0 +1,202 @@
+
+                                 Apache License
```
You're including code from the Go standard libraries, so I believe you would need to continue to ship with the original license:

https://golang.org/LICENSE: "Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer."

I have not checked word by word, but I think this is a 3-clause BSD license: https://en.wikipedia.org/wiki/BSD_licenses#3-clause_license_(%22BSD_License_2.0%22,_%22Revised_BSD_License%22,_%22New_BSD_License%22,_or_%22Modified_BSD_License%22)

bjackman replied:
Thanks, I've switched the license to the Go one. I'm awaiting advice from the licensing experts about whether anything further is needed to fix this.

bjackman commented:
Merging to unblock internal integration.Further review from @the80srobot  to be resolved in followup commits.
